libs/methodologies/bold/rule-processors/mass-id/src/uniqueness-check/uniqueness-check.helpers.ts

```diff
@@ -0,0 +1,223 @@
+import {
+  type ApiDocumentCheckDuplicatesResponse,
+  AuditApiService,
+} from '@carrot-fndn/shared/methodologies/audit-api';
+import {
+  type Document,
+  DocumentCategory,
+  type DocumentEvent,
+  DocumentEventActorType,
+  DocumentEventAttributeName,
+  DocumentEventMoveType,
+  DocumentEventName,
+  NewDocumentEventAttributeName,
+} from '@carrot-fndn/shared/methodologies/bold/types';
+import {
+  MethodologyDocumentEventLabel,
+  type NonEmptyString,
+} from '@carrot-fndn/shared/types';
+
+const { VEHICLE_LICENSE_PLATE } = NewDocumentEventAttributeName;
+const {
+  ACTOR_TYPE,
+  MOVE_TYPE,
+  VEHICLE_LICENSE_PLATE: OLD_VEHICLE_LICENSE_PLATE,
+} = DocumentEventAttributeName;
+const { ACTOR, DROP_OFF, MOVE, OPEN, PICK_UP } = DocumentEventName;
+const { DROP_OFF: DROP_OFF_MOVE_TYPE } = DocumentEventMoveType;
+const { RECYCLER, WASTE_GENERATOR } = MethodologyDocumentEventLabel;
+const { RECYCLER: RECYCLER_ACTOR_TYPE, SOURCE: SOURCE_ACTOR_TYPE } =
+  DocumentEventActorType;
+
+export type EventsData = {
+  dropOffEvent: DocumentEvent;
+  pickUpEvent: DocumentEvent;
+  recyclerEvent: DocumentEvent;
+  vehicleLicensePlate: NonEmptyString;
+  wasteGeneratorEvent: DocumentEvent;
+};
+
+export const createLicensePlateRegex = (
+  licensePlate: NonEmptyString,
+): string => {
+  const escapedValue = licensePlate
+    .replaceAll(/\s+/g, '')
+    .replaceAll(/[$()*+./?[\\\]^{|}-]/g, '\\$&');
+
+  return `^${escapedValue}$`;
+};
+
+/* istanbul ignore next */
+export const createAuditApiService = (): AuditApiService =>
+  new AuditApiService();
+
+export const mapMassIdV2Query = (
+  document: Document,
+  requiredData: EventsData,
+) => {
+  const {
+    dropOffEvent,
+    pickUpEvent,
+    recyclerEvent,
+    vehicleLicensePlate,
+    wasteGeneratorEvent,
+  } = requiredData;
+
+  const licensePlateRegex = createLicensePlateRegex(vehicleLicensePlate);
+
+  return {
+    match: {
+      $and: [
+        {
+          externalEvents: {
+            $elementMatch: {
+              'address.id': pickUpEvent.address.id,
+              externalCreatedAt: pickUpEvent.externalCreatedAt,
+              'metadata.attributes': {
+                $elemMatch: {
+                  name: VEHICLE_LICENSE_PLATE,
+                  value: { $options: 'i', $regex: licensePlateRegex },
+                },
+              },
+              name: PICK_UP,
+            },
+          },
+        },
+        {
+          externalEvents: {
+            $elementMatch: {
+              'address.id': dropOffEvent.address.id,
+              externalCreatedAt: dropOffEvent.externalCreatedAt,
```

Comment:
I'm wondering if a system error could generate the same massID with a few seconds or minutes away. It does not seem physically possible to have multiple masses with the same weight, generator and recycler in a short period, maybe 30 minutes. What do you think of checking a range of time in these externalCreatedAt matches?

Reply:
Having talked offline with you, I agree that it is too soon to implement something more flexible. We should look into this again in the future.